From 229767a384222eb5ebd0ba34f2634e4764a30725 Mon Sep 17 00:00:00 2001
From: Christophe Guibout <christophe.guibout@st.com>
Date: Mon, 23 Jun 2025 13:41:30 +0200
Subject: [PATCH 2/3] stm32mp: update env to support boot A/B

Patch to be reworked to avoid modifications in config_distro_bootcmd.h
---
 configs/stm32mp13_defconfig        | 2 +-
 configs/stm32mp15_defconfig        | 2 +-
 configs/stm32mp21_defconfig        | 2 +-
 configs/stm32mp23_defconfig        | 2 +-
 configs/stm32mp25_defconfig        | 2 +-
 include/config_distro_bootcmd.h    | 1 +
 include/configs/stm32mp13_common.h | 1 +
 include/configs/stm32mp15_common.h | 1 +
 include/configs/stm32mp21_common.h | 1 +
 include/configs/stm32mp23_common.h | 1 +
 include/configs/stm32mp25_common.h | 1 +
 11 files changed, 11 insertions(+), 5 deletions(-)

diff --git a/configs/stm32mp13_defconfig b/configs/stm32mp13_defconfig
index 830dfff653..43fb56be62 100644
--- a/configs/stm32mp13_defconfig
+++ b/configs/stm32mp13_defconfig
@@ -24,7 +24,7 @@ CONFIG_FIT_SIGNATURE=y
 CONFIG_LEGACY_IMAGE_FORMAT=y
 CONFIG_DISTRO_DEFAULTS=y
 CONFIG_BOOTDELAY=1
-CONFIG_BOOTCOMMAND="run bootcmd_stm32mp"
+CONFIG_BOOTCOMMAND="echo 'Boot A';setenv altboot 0;run bootcmd_stm32mp"
 CONFIG_FDT_SIMPLEFB=y
 CONFIG_SYS_PROMPT="STM32MP> "
 CONFIG_CMD_FWU_METADATA=y
diff --git a/configs/stm32mp15_defconfig b/configs/stm32mp15_defconfig
index 5821cb3d23..14861782bc 100644
--- a/configs/stm32mp15_defconfig
+++ b/configs/stm32mp15_defconfig
@@ -22,7 +22,7 @@ CONFIG_FIT_SIGNATURE=y
 CONFIG_LEGACY_IMAGE_FORMAT=y
 CONFIG_DISTRO_DEFAULTS=y
 CONFIG_BOOTDELAY=1
-CONFIG_BOOTCOMMAND="run bootcmd_stm32mp"
+CONFIG_BOOTCOMMAND="echo 'Boot A';setenv altboot 0;run bootcmd_stm32mp"
 CONFIG_FDT_SIMPLEFB=y
 CONFIG_SYS_PROMPT="STM32MP> "
 CONFIG_SYS_PBSIZE=1050
diff --git a/configs/stm32mp21_defconfig b/configs/stm32mp21_defconfig
index a95320828a..3bced8d184 100644
--- a/configs/stm32mp21_defconfig
+++ b/configs/stm32mp21_defconfig
@@ -24,7 +24,7 @@ CONFIG_FIT_SIGNATURE=y
 CONFIG_LEGACY_IMAGE_FORMAT=y
 CONFIG_DISTRO_DEFAULTS=y
 CONFIG_BOOTDELAY=1
-CONFIG_BOOTCOMMAND="run bootcmd_stm32mp"
+CONFIG_BOOTCOMMAND="echo 'Boot A';setenv altboot 0;run bootcmd_stm32mp"
 CONFIG_FDT_SIMPLEFB=y
 CONFIG_SYS_PROMPT="STM32MP> "
 CONFIG_CMD_BDINFO_EXTRA=y
diff --git a/configs/stm32mp23_defconfig b/configs/stm32mp23_defconfig
index bf6d130902..e1d76c6a62 100644
--- a/configs/stm32mp23_defconfig
+++ b/configs/stm32mp23_defconfig
@@ -24,7 +24,7 @@ CONFIG_FIT_SIGNATURE=y
 CONFIG_LEGACY_IMAGE_FORMAT=y
 CONFIG_DISTRO_DEFAULTS=y
 CONFIG_BOOTDELAY=1
-CONFIG_BOOTCOMMAND="run bootcmd_stm32mp"
+CONFIG_BOOTCOMMAND="echo 'Boot A';setenv altboot 0;run bootcmd_stm32mp"
 CONFIG_FDT_SIMPLEFB=y
 CONFIG_SYS_PROMPT="STM32MP> "
 CONFIG_CMD_BDINFO_EXTRA=y
diff --git a/configs/stm32mp25_defconfig b/configs/stm32mp25_defconfig
index 5b8aab47b8..9cc5d3cd59 100644
--- a/configs/stm32mp25_defconfig
+++ b/configs/stm32mp25_defconfig
@@ -24,7 +24,7 @@ CONFIG_FIT_SIGNATURE=y
 CONFIG_LEGACY_IMAGE_FORMAT=y
 CONFIG_DISTRO_DEFAULTS=y
 CONFIG_BOOTDELAY=1
-CONFIG_BOOTCOMMAND="run bootcmd_stm32mp"
+CONFIG_BOOTCOMMAND="echo 'Boot A';setenv altboot 0;run bootcmd_stm32mp"
 CONFIG_FDT_SIMPLEFB=y
 CONFIG_SYS_PROMPT="STM32MP> "
 CONFIG_CMD_BDINFO_EXTRA=y
diff --git a/include/config_distro_bootcmd.h b/include/config_distro_bootcmd.h
index 2a136b96a6..a95f57141a 100644
--- a/include/config_distro_bootcmd.h
+++ b/include/config_distro_bootcmd.h
@@ -541,6 +541,7 @@
 		"part list ${devtype} ${devnum} -bootable devplist; "     \
 		"env exists devplist || setenv devplist 1; "              \
 		"for distro_bootpart in ${devplist}; do "                 \
+		"if test ${altboot} = 1;then setexpr distro_bootpart ${distro_bootpart} + 1; fi;" \
 			"if fstype ${devtype} "                           \
 					"${devnum}:${distro_bootpart} "   \
 					"bootfstype; then "               \
diff --git a/include/configs/stm32mp13_common.h b/include/configs/stm32mp13_common.h
index 09b4ebc088..b5ddfddd58 100644
--- a/include/configs/stm32mp13_common.h
+++ b/include/configs/stm32mp13_common.h
@@ -85,6 +85,7 @@
 
 #define STM32MP_EXTRA \
 	"env_check=if env info -p -d -q; then env save; fi\0" \
+	"altbootcmd=echo 'Boot B';setenv altboot 1;run bootcmd_stm32mp\0" \
 	"boot_net_usb_start=true\0"
 
 #ifndef STM32MP_BOARD_EXTRA_ENV
diff --git a/include/configs/stm32mp15_common.h b/include/configs/stm32mp15_common.h
index 19b989cba2..f8d6bb83e3 100644
--- a/include/configs/stm32mp15_common.h
+++ b/include/configs/stm32mp15_common.h
@@ -90,6 +90,7 @@
 
 #define STM32MP_EXTRA \
 	"env_check=if env info -p -d -q; then env save; fi\0" \
+	"altbootcmd=echo 'Boot B';setenv altboot 1;run bootcmd_stm32mp\0" \
 	"boot_net_usb_start=true\0"
 
 #ifndef STM32MP_BOARD_EXTRA_ENV
diff --git a/include/configs/stm32mp21_common.h b/include/configs/stm32mp21_common.h
index 1b0daf4b5b..92a616f553 100644
--- a/include/configs/stm32mp21_common.h
+++ b/include/configs/stm32mp21_common.h
@@ -92,6 +92,7 @@
 
 #define STM32MP_EXTRA \
 	"env_check=if env info -p -d -q; then env save; fi\0" \
+	"altbootcmd=echo 'Boot B';setenv altboot 1;run bootcmd_stm32mp\0" \
 	"boot_net_usb_start=true\0"
 /*
  * memory layout for 96MB uncompressed/compressed kernel,
diff --git a/include/configs/stm32mp23_common.h b/include/configs/stm32mp23_common.h
index d9cb8b2392..f1fbbe6d4f 100644
--- a/include/configs/stm32mp23_common.h
+++ b/include/configs/stm32mp23_common.h
@@ -92,6 +92,7 @@
 
 #define STM32MP_EXTRA \
 	"env_check=if env info -p -d -q; then env save; fi\0" \
+	"altbootcmd=echo 'Boot B';setenv altboot 1;run bootcmd_stm32mp\0" \
 	"boot_net_usb_start=true\0"
 /*
  * memory layout for 96MB uncompressed/compressed kernel,
diff --git a/include/configs/stm32mp25_common.h b/include/configs/stm32mp25_common.h
index 711b82430e..cbe9514caf 100644
--- a/include/configs/stm32mp25_common.h
+++ b/include/configs/stm32mp25_common.h
@@ -92,6 +92,7 @@
 
 #define STM32MP_EXTRA \
 	"env_check=if env info -p -d -q; then env save; fi\0" \
+	"altbootcmd=echo 'Boot B';setenv altboot 1;run bootcmd_stm32mp\0" \
 	"boot_net_usb_start=true\0"
 /*
  * memory layout for 96MB uncompressed/compressed kernel,
-- 
2.34.1

